lib/elixir/src/elixir_module.erl

```diff
@@ -322,6 +323,35 @@ spec_for_macro({type, Line, 'fun', [{type, _, product, Args} | T]}) ->
 
 spec_for_macro(Else) -> Else.
 
+optional_callbacks_attributes([], _Specs) ->
+  [];
+optional_callbacks_attributes(OptionalCallbacksTypespec, Specs) ->
+  % We only take the specs of the callbacks (which are both callbacks and
+  % macrocallbacks).
+  Callbacks = [NameArity || {{callback, NameArity, _Spec}, _Line} <- Specs],
+
+  % We get all the optional callbacks as {Name, Arity} tuples alongside the line
+  % we'll use for this attribute.
+  OptionalCallbacks =
+    [{NameArity, ?m(E, line)} ||
+      {optional_callbacks, NamesArities, E} <- OptionalCallbacksTypespec,
+      NameArity <- NamesArities],
+  {NamesArities0, Lines} = lists:unzip(OptionalCallbacks),
+  Line = lists:min(Lines),
+
+  % We replace functions in the list of optional callbacks with their macro name
+  % if they're listed as macrocallbacks instead of regular callbacks.
+  NamesArities1 =
+    lists:map(fun({Name, Arity} = NameArity) ->
+                  MacroNameArity = {elixir_utils:macro_name(Name), Arity + 1},
+                  case lists:member(MacroNameArity, Callbacks) of
+                    true -> MacroNameArity;
+                    false -> NameArity
+                  end
+              end, NamesArities0),
+
+  [{attribute, Line, optional_callbacks, NamesArities1}].
```

Comment:
You can actually emit multiple of those, so that's what I would do. Emit one per line.